db/Pos/DB/Class.hs

```diff
@@ -99,19 +106,47 @@ dbTagToLens GStateDB     = gStateDB
 dbTagToLens LrcDB        = lrcDB
 dbTagToLens MiscDB       = miscDB
 
+-- | Key-value type family encapsulating the iterator (something we
+-- can iterate on) functionality.
+class DBIteratorClass i where
+    type IterKey   i :: *
+    type IterValue i :: *
+    iterKeyPrefix :: ByteString
+
+type IterType i = (IterKey i, IterValue i)
+
 -- | Pure read-only interface to the database.
--- TODO: add iteration, maybe something else.
-class MonadThrow m => MonadDBRead m where
+class (MonadBaseControl IO m, MonadThrow m) => MonadDBRead m where
     -- | This function takes tag and key and reads value associated
     -- with given key from DB corresponding to given tag.
     dbGet :: DBTag -> ByteString -> m (Maybe ByteString)
 
+    -- | Source producing iteration over given 'i'.
+    dbIterSource ::
+        ( DBIteratorClass i
+        , Bi (IterKey i)
+        , Bi (IterValue i)
+        ) => DBTag -> Proxy i -> Source (ResourceT m) (IterType i)
```

Comment:
It's a little strange that you deleted `Proxy` from `iterKeyPrefix` but are using it here.